Fail early on 401 from Maven Server / Avoid locking accounts

### Context
In our company, we have lots of gradle-based projects and thousands of users. We have centralized, ldap based identity management solutions. Our Nexus Server uses this central infrastructure to verify credentials and authenticate requests. We are also required to change passwords regularly.

When someone changes his/her password but forgets to update it in one of the `gradle.properties` files and launches a Gradle build that uses our Nexus repository, this causes his/her account to be locked immediately because Gradle will repeatedly send requests with the wrong credentials to the repository. This causes major disruption in our workflow as the global central account has to be unlocked by an international support team before work can continue.

### Expected Behavior
If a maven repository is used with credentials and Gradle gets at least one 401 response from it, Gradle should stop sending requests to this repository for the rest of the build and show that there's something wrong with the credentials.

### Current Behavior
Gradle continues dependency resolution and sending requests (10 in my case) with the wrong credentials to this Maven/Nexus server, causing the user's global ldap account to be instantly locked because of repeatedly failed authentication attempts. Here's my sample output:

```
FAILURE: Build failed with an exception.
 
* What went wrong:
A problem occurred configuring project ':my-project'.
> Could not resolve all files for configuration ':my-project:classpath'.
   > Could not download spring-boot-gradle-plugin-2.5.6.jar (org.springframework.boot:spring-boot-gradle-plugin:2.5.6)
      > Could not get resource 'https://nexus.example.com/repository/my-group/org/springframework/boot/spring-boot-gradle-plugin/2.5.6/spring-boot-gradle-plugin-2.5.6.jar'.
         > Could not GET 'https://nexus.example.com/repository/my-group/org/springframework/boot/spring-boot-gradle-plugin/2.5.6/spring-boot-gradle-plugin-2.5.6.jar'. Received status code 401 from server: Unauthorized
   > Could not download spring-boot-buildpack-platform-2.5.6.jar (org.springframework.boot:spring-boot-buildpack-platform:2.5.6)
      > Could not get resource 'https://nexus.example.com/repository/my-group/org/springframework/boot/spring-boot-buildpack-platform/2.5.6/spring-boot-buildpack-platform-2.5.6.jar'.
         > Could not GET 'https://nexus.example.com/repository/my-group/org/springframework/boot/spring-boot-buildpack-platform/2.5.6/spring-boot-buildpack-platform-2.5.6.jar'. Received status code 401 from server: Unauthorized
   > Could not download spring-boot-loader-tools-2.5.6.jar (org.springframework.boot:spring-boot-loader-tools:2.5.6)
      > Could not get resource 'https://nexus.example.com/repository/my-group/org/springframework/boot/spring-boot-loader-tools/2.5.6/spring-boot-loader-tools-2.5.6.jar'.
         > Could not GET 'https://nexus.example.com/repository/my-group/org/springframework/boot/spring-boot-loader-tools/2.5.6/spring-boot-loader-tools-2.5.6.jar'. Received status code 401 from server: Unauthorized
   > Could not download commons-compress-1.20.jar (org.apache.commons:commons-compress:1.20)
      > Could not get resource 'https://nexus.example.com/repository/my-group/org/apache/commons/commons-compress/1.20/commons-compress-1.20.jar'.
         > Could not GET 'https://nexus.example.com/repository/my-group/org/apache/commons/commons-compress/1.20/commons-compress-1.20.jar'. Received status code 401 from server: Unauthorized
   > Could not download spring-core-5.3.12.jar (org.springframework:spring-core:5.3.12)
      > Could not get resource 'https://nexus.example.com/repository/my-group/org/springframework/spring-core/5.3.12/spring-core-5.3.12.jar'.
         > Could not GET 'https://nexus.example.com/repository/my-group/org/springframework/spring-core/5.3.12/spring-core-5.3.12.jar'. Received status code 401 from server: Unauthorized
   > Could not download jackson-annotations-2.12.5.jar (com.fasterxml.jackson.core:jackson-annotations:2.12.5)
      > Could not get resource 'https://nexus.example.com/repository/my-group/com/fasterxml/jackson/core/jackson-annotations/2.12.5/jackson-annotations-2.12.5.jar'.
         > Could not GET 'https://nexus.example.com/repository/my-group/com/fasterxml/jackson/core/jackson-annotations/2.12.5/jackson-annotations-2.12.5.jar'. Received status code 401 from server: Unauthorized
   > Could not download jackson-core-2.12.5.jar (com.fasterxml.jackson.core:jackson-core:2.12.5)
      > Could not get resource 'https://nexus.example.com/repository/my-group/com/fasterxml/jackson/core/jackson-core/2.12.5/jackson-core-2.12.5.jar'.
         > Could not GET 'https://nexus.example.com/repository/my-group/com/fasterxml/jackson/core/jackson-core/2.12.5/jackson-core-2.12.5.jar'. Received status code 401 from server: Unauthorized
   > Could not download jackson-module-parameter-names-2.12.5.jar (com.fasterxml.jackson.module:jackson-module-parameter-names:2.12.5)
      > Could not get resource 'https://nexus.example.com/repository/my-group/com/fasterxml/jackson/module/jackson-module-parameter-names/2.12.5/jackson-module-parameter-names-2.12.5.jar'.
         > Could not GET 'https://nexus.example.com/repository/my-group/com/fasterxml/jackson/module/jackson-module-parameter-names/2.12.5/jackson-module-parameter-names-2.12.5.jar'. Received status code 401 from server: Unauthorized
   > Could not download jackson-databind-2.12.5.jar (com.fasterxml.jackson.core:jackson-databind:2.12.5)
      > Could not get resource 'https://nexus.example.com/repository/my-group/com/fasterxml/jackson/core/jackson-databind/2.12.5/jackson-databind-2.12.5.jar'.
         > Could not GET 'https://nexus.example.com/repository/my-group/com/fasterxml/jackson/core/jackson-databind/2.12.5/jackson-databind-2.12.5.jar'. Received status code 401 from server: Unauthorized
   > Could not download spring-jcl-5.3.12.jar (org.springframework:spring-jcl:5.3.12)
      > Could not get resource 'https://nexus.example.com/repository/my-group/org/springframework/spring-jcl/5.3.12/spring-jcl-5.3.12.jar'.
         > Could not GET 'https://nexus.example.com/repository/my-group/org/springframework/spring-jcl/5.3.12/spring-jcl-5.3.12.jar'. Received status code 401 from server: Unauthorized
```

--- a/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/ErrorHandlingModuleComponentRepository.java
+++ b/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/ErrorHandlingModuleComponentRepository.java
@@ -244,7 +244,7 @@ public class ErrorHandlingModuleComponentRepository implements ModuleComponentRe
                     unexpectedFailure = throwable;
                     failure = onError.transform(throwable);
                 }
-                boolean doNotRetry = !NetworkingIssueVerifier.isLikelyTransientNetworkingIssue(failure);
+                boolean doNotRetry = NetworkingIssueVerifier.isLikelyPermanentNetworkIssue(failure) || !NetworkingIssueVerifier.isLikelyTransientNetworkingIssue(failure);
                 if (doNotRetry || retries == maxTentativesCount) {
                     if (unexpectedFailure != null) {
                         repositoryBlacklister.disableRepository(repositoryId, unexpectedFailure);

--- a/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/repositories/transport/NetworkingIssueVerifier.java
+++ b/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/repositories/transport/NetworkingIssueVerifier.java
@@ -61,4 +61,23 @@ public class NetworkingIssueVerifier {
         return statusCode == HttpStatus.SC_REQUEST_TIMEOUT ||
             statusCode == 429; // Too many requests (not available through HttpStatus.XXX)
     }
+
+    public static <E extends Throwable> boolean isLikelyPermanentNetworkIssue(E failure) {
+        if (failure instanceof HttpErrorStatusCodeException) {
+            return isClientAuthenticationError(((HttpErrorStatusCodeException) failure).getStatusCode());
+        }
+        if (failure instanceof DefaultMultiCauseException) {
+            List<? extends Throwable> causes = ((DefaultMultiCauseException) failure).getCauses();
+            for (Throwable cause : causes) {
+                if (isLikelyPermanentNetworkIssue(cause)) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    private static boolean isClientAuthenticationError(int statusCode) {
+        return statusCode == HttpStatus.SC_UNAUTHORIZED || statusCode == HttpStatus.SC_FORBIDDEN;
+    }
 }

--- a/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/repositories/transport/NetworkingIssueVerifier.java
+++ b/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/repositories/transport/NetworkingIssueVerifier.java
@@ -26,6 +26,9 @@ import java.util.List;
 
 public class NetworkingIssueVerifier {
 
+    // Too many requests (not available through HttpStatus.XXX)
+    private static final int SC_TOO_MANY_REQUESTS = 429;
+
     /**
      * Determines if an error should cause a retry. We will currently retry:
      * <ul>
@@ -58,8 +61,7 @@ public class NetworkingIssueVerifier {
     }
 
     private static boolean isTransientClientError(int statusCode) {
-        return statusCode == HttpStatus.SC_REQUEST_TIMEOUT ||
-            statusCode == 429; // Too many requests (not available through HttpStatus.XXX)
+        return statusCode == HttpStatus.SC_REQUEST_TIMEOUT || statusCode == SC_TOO_MANY_REQUESTS;
     }
 
     public static <E extends Throwable> boolean isLikelyPermanentNetworkIssue(E failure) {

--- a/subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/ErrorHandlingModuleComponentRepositoryTest.groovy
+++ b/subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/ErrorHandlingModuleComponentRepositoryTest.groovy
@@ -65,7 +65,9 @@ class ErrorHandlingModuleComponentRepositoryTest extends Specification {
     @Shared
     def tooManyRequests = status(429)
     @Shared
-    def unauthorized = status(403)
+    def forbidden = status(403)
+    @Shared
+    def unauthorized = status(401)
 
     @Subject
     ErrorHandlingModuleComponentRepository.ErrorHandlingModuleComponentRepositoryAccess access
@@ -269,12 +271,47 @@ class ErrorHandlingModuleComponentRepositoryTest extends Specification {
         [maxRetries, exception, effectiveRetries] << retryCombinations()
     }
 
+    @Unroll
+    def "blocks repo immediately when receives #desc"() {
+        access = createAccess(1)
+
+        given:
+        def artifact = Mock(ComponentArtifactMetadata)
+        def artifactId = Mock(ComponentArtifactIdentifier)
+        def moduleSources = ImmutableModuleSources.of(Mock(ModuleSource))
+        def result = Mock(BuildableArtifactResolveResult)
+        artifact.getId() >> artifactId
+        delegate.resolveArtifact(artifact, moduleSources, result) >> { throw exception }
+        repositoryBlacklister.isDisabled(REPOSITORY_ID) >> false
+
+        when: 'repo is not disabled'
+        access.resolveArtifact(artifact, moduleSources, result)
+
+        then: 'resolution fails and repo is disabled'
+        1 * repositoryBlacklister.disableRepository(REPOSITORY_ID, { hasCause(it, exception) })
+        1 * result.failed(_ as ArtifactResolveException)
+
+        when: 'repo is already disabled'
+        repositoryBlacklister.isDisabled(REPOSITORY_ID) >> true
+        access.resolveArtifact(artifact, moduleSources, result)
+
+        then: 'resolution fails directly'
+        1 * result.failed(_ as ArtifactResolveException)
+        0 * delegate._
+
+        where:
+        desc | exception
+        "unauthorized (401)" | unauthorized
+        "forbidden (403)" | forbidden
+    }
+
     List<List<?>> retryCombinations() {
         def retries = []
         (1..3).each { ret ->
             // no retries on runtime errors, missing resources, authentication errors
             retries << [ret, runtimeError, 1]
             retries << [ret, missing, 1]
+            retries << [ret, forbidden, 1]
             retries << [ret, unauthorized, 1]
             // retries on connect timeouts, too many requests, client timeouts
             retries << [ret, connectTimeout, ret]

